### PR TITLE
Renew cert fix

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: smcontroller/smcontroller.go [hunk 1]
```diff
@@ -116,6 +132,9 @@ func New(
 		systemQuotaAlertChan:      make(chan cloudprotocol.SystemQuotaAlert, statusChanSize),
 		nodeConfigStatusChan:      make(chan unitconfig.NodeConfigStatus, statusChanSize),
 		nodes:                     make(map[string]*smHandler),
+		certChan:                  make(<-chan *iamanager.CertInfo),
```
Comment (al1img) [inline smcontroller/smcontroller.go:135]: Make is useless here.

File: smcontroller/smcontroller.go [hunk 2]
```diff
@@ -124,48 +143,38 @@ func New(
...
-		tlsConfig, err := cryptcoxontext.GetClientMutualTLSConfig(certURL, keyURL)
-		if err != nil {
+		if ch, err = controller.certProvider.SubscribeCertChanged(cfg.CertStorage); err != nil {
 			return nil, aoserrors.Wrap(err)
 		}
 
-		opts = append(opts, grpc.Creds(credentials.NewTLS(tlsConfig)))
-	} else {
-		log.Info("GRPC server starts in insecure mode")
+		controller.certChan = ch
```
Comment (al1img) [inline smcontroller/smcontroller.go:165]: You can pass cert channel as parameter instead of using private member.

File: cmserver/cmserver.go [hunk 3]
```diff
@@ -117,52 +128,35 @@ type CertificateProvider interface {
...
 	server = &CMServer{
+		config:        cfg,
+		certProvider:  certProvider,
+		cryptocontext: cryptocontext,
+		insecureConn:  insecure,
+
+		grpcServer:        grpchelpers.NewGRPCServer(cfg.CMServerURL),
 		currentFOTAStatus: handler.GetFOTAStatus(),
 		currentSOTAStatus: handler.GetSOTAStatus(),
+		certChan:          make(<-chan *iamanager.CertInfo),
```
Comment (al1img) [inline cmserver/cmserver.go:142]: Make is useless here. Rename to `certChannel` to be aligned with `stopChannel`.

File: smcontroller/smcontroller.go [hunk 4]
```diff
@@ -65,10 +73,12 @@ type Controller struct {
 	runInstancesStatusChan    chan launcher.NodeRunInstanceStatus
 	systemQuotaAlertChan      chan cloudprotocol.SystemQuotaAlert
 	nodeConfigStatusChan      chan unitconfig.NodeConfigStatus
+	certChan                  <-chan *iamanager.CertInfo
```
Comment (al1img) [inline smcontroller/smcontroller.go:76]: `certChannel`
`closeChannel`

File: iamclient/iamclient.go [hunk 5]
```diff
@@ -51,32 +47,56 @@ const (
...
+
 	identService        pb.IAMPublicIdentityServiceClient
 	certificateService  pb.IAMCertificateServiceClient
 	provisioningService pb.IAMProvisioningServiceClient
+	publicNodesService  pb.IAMPublicNodesServiceClient
+	nodesService        pb.IAMNodesServiceClient
+
+	nodeInfoSubs *nodeInfoChangeSub
+
+	tlsCertChan      <-chan *pb.CertInfo
```
Comment (al1img) [inline iamclient/iamclient.go:80]: certChannel

File: iamclient/iamclient.go [hunk 6]
```diff
@@ -51,32 +47,56 @@ const (
...
+
+	reconnectTimer *time.Timer
+}
 
-	closeChannel       chan struct{}
-	publicNodesService pb.IAMPublicNodesServiceClient
-	nodesService       pb.IAMNodesServiceClient
-	nodeInfoListeners  []chan cloudprotocol.NodeInfo
+// certSubscription generic subscription for IAM public messages.
+type iamSubscription[grpcStream *pb.IAMPublicService_SubscribeCertChangedClient |
```
Comment (al1img) [inline iamclient/iamclient.go:89]: Where is `IAMPublicService_SubscribeCertChangedClient` used?

File: iamclient/iamclient.go [hunk 7]
```diff
@@ -51,32 +47,56 @@ const (
...
 	certificateService  pb.IAMCertificateServiceClient
 	provisioningService pb.IAMProvisioningServiceClient
+	publicNodesService  pb.IAMPublicNodesServiceClient
+	nodesService        pb.IAMNodesServiceClient
+
+	nodeInfoSubs *nodeInfoChangeSub
+
+	tlsCertChan      <-chan *pb.CertInfo
+	closeChannel     chan struct{}
+	disableReconnect atomic.Bool
```
Comment (al1img) [inline iamclient/iamclient.go:82]: Use ordinal bool and guard with mutex. It is already guarded in most the cases.

File: iamclient/iamclient.go [hunk 8]
```diff
@@ -99,16 +119,25 @@ type CertificateProvider interface {
...
+		config:                 config,
+		sender:                 sender,
+		cryptocontext:          cryptocontext,
+		insecure:               insecure,
+		publicConnection:       grpchelpers.NewBlockingGRPCConn(),
+		nodeInfoSubs: &nodeInfoChangeSub{
+			listeners: make([]chan cloudprotocol.NodeInfo, 0),
+		},
+
+		tlsCertChan:      make(<-chan *pb.CertInfo),
```
Comment (al1img) [inline iamclient/iamclient.go:138]: make is useless

File: iamclient/iamclient_test.go [hunk 9]
```diff
@@ -347,6 +349,96 @@ func TestGetCertificates(t *testing.T) {
...
+
+	defer publicServer.close()
+	defer protectedServer.close()
+
+	client, err := iamclient.New(&config.Config{
+		IAMProtectedServerURL: protectedServerURL,
+		IAMPublicServerURL:    publicServerURL,
+	}, &testSender{}, nil, true)
+	if err != nil {
+		t.Fatalf("Can't create IAM client: %s", err)
```
Comment (al1img) [inline iamclient/iamclient_test.go:366]: %v

File: iamclient/iamclient_test.go [hunk 10]
```diff
@@ -347,6 +349,96 @@ func TestGetCertificates(t *testing.T) {
...
+
+	iamCert := pb.CertInfo{
+		Type:    "iam",
+		CertUrl: "iam cert url",
+		KeyUrl:  "iam key url",
+	}
+
+	ch, err := client.SubscribeCertChanged("iam")
+	if err != nil {
+		t.Fatalf("Can't subscribe cert change err=%v", err)
```
Comment (al1img) [inline iamclient/iamclient_test.go:378]: Use one pattern everywhere: either `"Message: %v"` or `"Message: err=%v"`

File: cmserver/cmserver.go [hunk 11]
```diff
@@ -317,6 +310,46 @@ func (server *CMServer) notifyAllClients(notification *pb.SchedulerNotifications
...
+	if server.restartTimer != nil {
+		server.restartTimer.Stop()
+		server.restartTimer = nil
+	}
+
+	if err := server.startGRPCServer(); err != nil {
+		log.WithField("err", err).Error("CMServer failed to start GRPC server")
+
+		server.restartTimer = time.AfterFunc(cmRestartInterval, func() {
+			server.restartTimer = nil
```
Comment (al1img) [inline cmserver/cmserver.go:347]: Should be guarded with mutex.

File: umcontroller/umcontroller.go [hunk 12]
```diff
@@ -49,14 +52,21 @@ import (
...
+	insecure      bool
+
+	server *umCtrlServer
 
 	eventChannel     chan umCtrlInternalMsg
 	nodeInfoProvider NodeInfoProvider
 	nodeInfoChannel  <-chan cloudprotocol.NodeInfo
-	stopChannel      chan bool
-	componentDir     string
+	certChan         <-chan *iamanager.CertInfo
```
Comment (al1img) [inline umcontroller/umcontroller.go:66]: `certChannel`

File: umcontroller/umcontroller.go [hunk 13]
```diff
@@ -225,10 +247,16 @@ func New(config *config.Config, storage storage, certProvider CertificateProvide
...
+		config:        config,
+		storage:       storage,
+		certProvider:  certProvider,
+		cryptocontext: cryptocontext,
+		insecure:      insecure,
+
 		eventChannel:         make(chan umCtrlInternalMsg),
 		nodeInfoProvider:     nodeInfoProvider,
 		nodeInfoChannel:      nodeInfoProvider.SubscribeNodeInfoChange(),
+		certChan:             make(<-chan *iamanager.CertInfo),
```
Comment (al1img) [inline umcontroller/umcontroller.go:259]: remove make

File: iamclient/iamclient.go [hunk 14]
```diff
@@ -117,35 +146,29 @@ func New(
...
-	localClient.identService = pb.NewIAMPublicIdentityServiceClient(localClient.publicConnection)
-	localClient.publicNodesService = pb.NewIAMPublicNodesServiceClient(localClient.publicConnection)
-
-	if localClient.protectedConnection, err = localClient.createProtectedConnection(
-		config, cryptocontext, insecure); err != nil {
-		return nil, err
+	if !insecure {
+		if ch, err := localClient.SubscribeCertChanged(config.CertStorage); err != nil {
+			return nil, aoserrors.Wrap(err)
+		} else {
```
Comment (al1img) [inline iamclient/iamclient.go:156]: No need `else` after return.

File: iamclient/iamclient.go [hunk 15]
```diff
@@ -474,102 +512,128 @@ func (client *Client) ResumeNode(nodeID string) error {
...
+			client.Lock()
+			client.reconnect()
+			client.Unlock()
 		}
+	}
+}
 
-		secureOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+func (client *Client) reconnect() {
+	if client.disableReconnect.CompareAndSwap(false, true) {
```
Comment (al1img) [inline iamclient/iamclient.go:615]: It is unclear why disableReconnect is required. Need to discuss.

File: smcontroller/smcontroller.go [hunk 16]
```diff
@@ -554,3 +538,57 @@ func (controller *Controller) sendConnectionStatus() {
...
+	controller.Lock()
+	defer controller.Unlock()
+
+	if controller.restartTimer != nil {
+		controller.restartTimer.Stop()
+		controller.restartTimer = nil
+	}
+
+	if err := controller.startGRPCServer(); err != nil {
+		log.WithField("err", err).Error("SMController failed to start GRPC server")
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply smcontroller/smcontroller.go:575]: fixed

File: iamclient/iamclient.go [hunk 17]
```diff
@@ -160,13 +183,11 @@ func (client *Client) GetSystemID() (systemID string) {
...
 }
 
-// GetCurrentNodeInfo returns info for current node.
-func (client *Client) GetCurrentNodeInfo() (nodeInfo cloudprotocol.NodeInfo, err error) {
-	return client.GetNodeInfo(client.GetNodeID())
-}
-
 // GetNodeInfo returns node info.
 func (client *Client) GetNodeInfo(nodeID string) (nodeInfo cloudprotocol.NodeInfo, err error) {
+	client.Lock()
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go:201]: to synchronize with reconnection logic that changes
client.publicNodesService instance

File: smcontroller/smcontroller.go (hunk 1 above)
Comment (mykola-kobets-epam) [reply smcontroller/smcontroller.go:135]: make initializes certChannel  for server.handleChannels() goroutine for insecure connection case

File: smcontroller/smcontroller.go [hunk 18]
```diff
@@ -124,48 +143,38 @@ func New(
...
 	if !insecureConn {
-		certURL, keyURL, err := certProvider.GetCertificate(cfg.CertStorage, nil, "")
-		if err != nil {
-			return nil, aoserrors.Wrap(err)
-		}
+		var ch <-chan *iamanager.CertInfo
 
-		tlsConfig, err := cryptcoxontext.GetClientMutualTLSConfig(certURL, keyURL)
-		if err != nil {
+		if ch, err = controller.certProvider.SubscribeCertChanged(cfg.CertStorage); err != nil {
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply smcontroller/smcontroller.go:161]: ch is used outside if block

File: smcontroller/smcontroller.go (hunk 2 above)
Comment (mykola-kobets-epam) [reply smcontroller/smcontroller.go:165]: done

File: cmserver/cmserver.go (hunk 3 above)
Comment (mykola-kobets-epam) [reply cmserver/cmserver.go:142]: done

File: smcontroller/smcontroller.go (hunk 4 above)
Comment (mykola-kobets-epam) [reply smcontroller/smcontroller.go:76]: done

File: iamclient/iamclient.go (hunk 6 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go:95]: fixed in aos_common

File: iamclient/iamclient.go (hunk 7 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: it is required to skip reconnecting when it is in process. atomic fits better for this particular case

File: iamclient/iamclient.go (hunk 8 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: removed

File: iamclient/iamclient_test.go (hunk 10 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient_test.go]: done

File: iamclient/iamclient_test.go (hunk 9 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient_test.go]: done

File: cmserver/cmserver.go (hunk 11 above)
Comment (mykola-kobets-epam) [reply cmserver/cmserver.go:347]: server.restartTimer = nil is not needed

File: umcontroller/umcontroller.go (hunk 12 above)
Comment (mykola-kobets-epam) [reply umcontroller/umcontroller.go:66]: done

File: umcontroller/umcontroller.go (hunk 13 above)
Comment (mykola-kobets-epam) [reply umcontroller/umcontroller.go:259]: initializes certChannel for server.handleChannels() goroutine for insecure case

File: iamclient/iamclient.go (hunk 14 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go:169]: done

File: iamclient/iamclient.go (hunk 15 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: disable reconnection attempts if any in progress

File: umcontroller/umcontroller.go [hunk 19]
```diff
@@ -29,6 +29,8 @@ import (
 	"sync"
 	"time"
 
+	"golang.org/x/exp/slices"
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply umcontroller/umcontroller.go:32]: done

File: smcontroller/smcontroller.go [hunk 20]
```diff
@@ -124,48 +141,37 @@ func New(
...
 	return controller, nil
 }
 
 // Close closes SM controller.
 func (controller *Controller) Close() error {
 	log.Debug("Close SM controller")
 
-	controller.stopServer()
+	controller.closeChannel <- struct{}{}
+	controller.stopGRPCServer()
```
Comment (al1img) [inline smcontroller/smcontroller.go:174]: Should stopGRPCServer be guarded with mutex to do not conflict with reconnect?

Comment (mykola-kobets-epam) [reply smcontroller/smcontroller.go:174]: It uses GRPCServer.stop from grpchelpers that is already guarded

**Review by MykolaSolyanko — APPROVED** [review]

**Review by al1img — APPROVED** [review]

**Pull request merged**